libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "AC_SafeRTL.h"
+
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo AC_SafeRTL::var_info[] = {
+    // @Param: ACCURACY
+    // @DisplayName: SafeRTL _accuracy
+    // @Description: SafeRTL _accuracy. The distance flown (in meters) before a new point is dropped.
+    // @Range: 0 10
+    // @User: Advanced
+    AP_GROUPINFO("ACCURACY", 0, AC_SafeRTL, _accuracy, SAFERTL_ACCURACY_DEFAULT),
+
+    // @Param: PATH_LEN
+    // @DisplayName: SafeRTL path length
+    // @Description: SafeRTL path length. The maximum number of points saved by SafeRTL. Set to 0 to disable SafeRTL.
+    // @Range: 0 500
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("PATH_LEN", 1, AC_SafeRTL, _desired_path_len, SAFERTL_PATH_LEN_DEFAULT),
+
+    AP_GROUPEND
+};
+
+/*
+*    This library is used for Copter's Safe Return-to-Launch feature. It stores
+*    "breadcrumbs" in memory, up to a certain number  of breadcrumbs have been stored,
+*    and then cleans up those breadcrumbs when space is filling up. When Safe-RTL is
+*    triggered, a more thorough cleanup occurs, and then the resulting path can
+*    is fed into WP_NAV to fly it.
+*
+*    The cleanup consists of two parts, pruning and simplification. Pruning works
+*    by calculating the closest distance reached by the line segment between any
+*    two pairs of sequential points, and cuts out anything between two points
+*    when their line segments get close. This algorithm will never compare two
+*    consecutive line segments. Obviously the segments (p1,p2) and (p2,p3) will
+*    get very close (they touch), but there would be nothing to trim between them.
+*    The simplification step uses the Ramer-Douglas-Peucker algorithm.
+*    See Wikipedia for description.
+*
+*    The simplification and pruning algorithms do not alter the path in memory,
+*    and are designed to run in the background. They each have a parameter that
+*    decides how long they will run before saving their state and returning. They
+*    are both anytime algorithms, which is helpful when memory is filling up and
+*    we just need to quickly identify a handful of points which can be deleted.
+*    The algorithms can also report if they are done. If not, the thorough cleanup
```

Comment:
i've tried to improve the wording in a follow up commit.  I met let @squilter to try and improve it further.